File: swift/Sources/Rezel/Grammars/JavaScriptGrammar.swift
import Foundation

private let space: [Int] = [9, 10, 11, 12, 13, 32, 133, 160, 5760, 8192, 8193, 8194, 8195, 8196, 8197, 8198, 8199, 8200,
                            8201, 8202, 8232, 8233, 8239, 8287, 12288]

private let braceR = 125
private let semicolon = 59
private let slash = 47
private let star = 42
private let plus = 43
private let minus = 45
private let lt = 60
private let comma = 44
private let question = 63
private let dot = 46
private let bracketL = 91

nonisolated(unsafe) let jsTrackNewline = ContextTracker(
	start: false as Any,
	shift: { context, term, stack, _ in
		let name = stack.parser.termNames?[term] ?? ""
		if name == "LineComment" || name == "BlockComment" || name == "spaces" {
			return context
		}
		return (name == "newline") as Any
	},
	strict: false
)

func makeJsExternalTokenizer(name: String, terms: [String: Int]) -> TokenizerProtocol {
	if name == "insertSemicolon" {
		let insertSemi = terms["insertSemi"]!
		return ExternalTokenizer({ input, stack in
			let next = input.next
			if next == braceR || next == -1 || (stack.context as? Bool == true) {
				input.acceptToken(insertSemi)
			}
		}, contextual: true, fallback: true)
	}

	if name == "noSemicolon" {
		let noSemi = terms["noSemi"]!
		return ExternalTokenizer({ input, stack in
			let next = input.next
			if space.contains(next) { return }
			if next == slash {
				let after = input.peek(1)
				if after == slash || after == star { return }
			}
			if next != braceR, next != semicolon, next != -1, !(stack.context as? Bool == true) {
				input.acceptToken(noSemi)
			}
		}, contextual: true)
	}

	if name == "noSemicolonType" {
		let noSemiType = terms["noSemiType"]!
		return ExternalTokenizer({ input, stack in
			if input.next == bracketL, !(stack.context as? Bool == true) {
				input.acceptToken(noSemiType)
			}
		}, contextual: true)
	}

	if name == "operatorToken" {
		let incdec = terms["incdec"]!
		let incdecPrefix = terms["incdecPrefix"]!
		let questionDot = terms["questionDot"]!
		return ExternalTokenizer({ input, stack in
			let next = input.next
			if next == plus || next == minus {
				input.advance()
				if next == input.next {
					input.advance()
					let mayPostfix = !(stack.context as? Bool == true) && stack.canShift(incdec)
					input.acceptToken(mayPostfix ? incdec : incdecPrefix)
				}
			} else if next == question, input.peek(1) == dot {
				input.advance()
				input.advance()
				if input.next < 48 || input.next > 57 {
					input.acceptToken(questionDot)
				}
			}
		}, contextual: true)
	}

	if name == "jsx" {
		let JSXStartTag = terms["JSXStartTag"]!
		return ExternalTokenizer({ input, stack in
			if input.next != lt { return }
			let parser = stack.parser
			let keys = Array(parser.dialects.keys)
			let jsxIdx = keys.firstIndex(of: "jsx")
			if let jsxIdx = jsxIdx {
				if !stack.dialectEnabled(jsxIdx) { return }
			} else {
				return
			}
			input.advance()
			if input.next == slash { return }
			var back = 0
			while space.contains(input.next) {
				input.advance(); back += 1
			}
			if jsIdentifierChar(input.next, true) {
				input.advance()
				back += 1
				while jsIdentifierChar(input.next, false) {
					input.advance(); back += 1
				}
				while space.contains(input.next) {
					input.advance(); back += 1
				}
				if input.next == comma { return }
				let extendsStr = "extends"
				let extendsScalars = Array(extendsStr.unicodeScalars)
				for i in 0 ... extendsStr.count {
					if i == extendsStr.count {
						if !jsIdentifierChar(input.next, true) { return }
						break
					}
					if input.next != Int(extendsScalars[i].value) { break }
					input.advance()
					back += 1
				}
			}
			input.acceptToken(JSXStartTag, endOffset: -back)
		}, contextual: true)
	}

	fatalError("Unknown JS external tokenizer: \(name)")
}

private func jsIdentifierChar(_ ch: Int, _ start: Bool) -> Bool {
	return (ch >= 65 && ch <= 90) || (ch >= 97 && ch <= 122) || ch == 95 || ch >= 192 ||
		(!start && ch >= 48 && ch <= 57)
}

private nonisolated(unsafe) let definition = hlTags["definition"] as! (Tag) -> Tag
private nonisolated(unsafe) let functionMod = hlTags["function"] as! (Tag) -> Tag
private nonisolated(unsafe) let specialMod = hlTags["special"] as! (Tag) -> Tag
private nonisolated(unsafe) let standardMod = hlTags["standard"] as! (Tag) -> Tag

nonisolated(unsafe) let jsHighlighting = styleTags([
	"get set async static": hlTags["modifier"] as Any,
	"for while do if else switch try catch finally return throw break continue default case defer": hlTags["controlKeyword"] as Any,
	"in of await yield void typeof delete instanceof as satisfies": hlTags["operatorKeyword"] as Any,
	"let var const using function class extends": hlTags["definitionKeyword"] as Any,
	"import export from": hlTags["moduleKeyword"] as Any,
	"with debugger new": hlKeyword,
	"TemplateString": specialMod(hlString),
	"super": hlTags["atom"] as Any,
	"BooleanLiteral": hlTags["bool"] as Any,
	"this": hlTags["self"] as Any,
	"null": hlTags["null"] as Any,
	"Star": hlTags["modifier"] as Any,
	"VariableName": hlTags["variableName"] as Any,
	"CallExpression/VariableName TaggedTemplateExpression/VariableName": functionMod(hlTags["variableName"] as! Tag),
	"VariableDefinition": definition(hlTags["variableName"] as! Tag),
	"Label": hlTags["labelName"] as Any,
	"PropertyName": hlPropertyName,
	"PrivatePropertyName": specialMod(hlPropertyName),
	"CallExpression/MemberExpression/PropertyName": functionMod(hlPropertyName),
	"FunctionDeclaration/VariableDefinition": functionMod(definition(hlTags["variableName"] as! Tag)),
	"ClassDeclaration/VariableDefinition": definition(hlTags["className"] as! Tag),
	"NewExpression/VariableName": hlTags["className"] as Any,
	"PropertyDefinition": definition(hlPropertyName),
	"PrivatePropertyDefinition": definition(specialMod(hlPropertyName)),
	"UpdateOp": hlTags["updateOperator"] as Any,
	"LineComment Hashbang": hlTags["lineComment"] as Any,
	"BlockComment": hlTags["blockComment"] as Any,
	"Number": hlNumber,
	"String": hlString,
	"Escape": hlTags["escape"] as Any,
	"ArithOp": hlTags["arithmeticOperator"] as Any,
	"LogicOp": hlTags["logicOperator"] as Any,
	"BitOp": hlTags["bitwiseOperator"] as Any,
	"CompareOp": hlTags["compareOperator"] as Any,
	"RegExp": hlTags["regexp"] as Any,
	"Equals": hlTags["definitionOperator"] as Any,
	"Arrow": functionMod(hlPunctuation),
	": Spread": hlPunctuation,
	"( )": hlTags["paren"] as Any,
	"[ ]": hlTags["squareBracket"] as Any,
	"{ }": hlTags["brace"] as Any,
	"InterpolationStart InterpolationEnd": specialMod(hlTags["brace"] as! Tag),
	".": hlTags["derefOperator"] as Any,
	", ;": hlTags["separator"] as Any,
	"@": hlMeta,
	"TypeName": hlTypeName,
	"TypeDefinition": definition(hlTypeName),
	"type enum interface implements namespace module declare": hlTags["definitionKeyword"] as Any,
	"abstract global Privacy readonly override": hlTags["modifier"] as Any,
	"is keyof unique infer asserts": hlTags["operatorKeyword"] as Any,
	"JSXAttributeValue": hlTags["attributeValue"] as Any,
	"JSXText": hlContent,
	"JSXStartTag JSXStartCloseTag JSXSelfCloseEndTag JSXEndTag": hlTags["angleBracket"] as Any,
	"JSXIdentifier JSXNamespacedName": hlTags["tagName"] as Any,
	"JSXAttribute/JSXIdentifier JSXAttribute/JSXNamespacedName": hlTags["attributeName"] as Any,
	"JSXBuiltin/JSXIdentifier": standardMod(hlTags["tagName"] as! Tag),
])

public nonisolated(unsafe) let javaScriptParser: LRParser = {
	let spec = JavaScriptParserData.makeSpec(
		externals: [
			"insertSemicolon": makeJsExternalTokenizer(name: "insertSemicolon", terms: JavaScriptParserData.termTable),
			"noSemicolon": makeJsExternalTokenizer(name: "noSemicolon", terms: JavaScriptParserData.termTable),
			"noSemicolonType": makeJsExternalTokenizer(name: "noSemicolonType", terms: JavaScriptParserData.termTable),
			"operatorToken": makeJsExternalTokenizer(name: "operatorToken", terms: JavaScriptParserData.termTable),
			"jsx": makeJsExternalTokenizer(name: "jsx", terms: JavaScriptParserData.termTable),
		],
		propSources: [jsHighlighting],
		context: jsTrackNewline
	)
	return LRParser(spec: spec)
}()

File: dotnet/Rezel/Grammars/JavaScriptGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class JavaScriptGrammar
{
    private static readonly HashSet<int> Space = [
        9, 10, 11, 12, 13, 32, 133, 160, 5760, 8192, 8193, 8194, 8195, 8196, 8197, 8198, 8199, 8200,
        8201, 8202, 8232, 8233, 8239, 8287, 12288
    ];

    private const int BraceR = 125;
    private const int Semicolon = 59;
    private const int Slash = 47;
    private const int Star = 42;
    private const int Plus = 43;
    private const int Minus = 45;
    private const int Lt = 60;
    private const int Comma = 44;
    private const int Question = 63;
    private const int Dot = 46;
    private const int BracketL = 91;

    private static bool JsIdentifierChar(int ch, bool start)
    {
        return (ch >= 65 && ch <= 90) || (ch >= 97 && ch <= 122) || ch == 95 || ch >= 192 ||
            (!start && ch >= 48 && ch <= 57);
    }

    public static readonly ContextTracker JsTrackNewline = new(
        start: (object?)false,
        shift: (context, term, stack, _) =>
        {
            var name = stack.Parser.TermNames?.GetValueOrDefault(term) ?? "";
            if (name is "LineComment" or "BlockComment" or "spaces")
                return context;
            return name == "newline";
        },
        @strict: false
    );

    private static ITokenizer MakeJsExternalTokenizer(string name, Dictionary<string, int> terms)
    {
        if (name == "insertSemicolon")
        {
            var insertSemi = terms["insertSemi"];
            return new ExternalTokenizer((input, stack) =>
            {
                var next = input.Next;
                if (next == BraceR || next == -1 || ((bool?)stack.Context == true))
                    input.AcceptToken(insertSemi);
            }, contextual: true, fallback: true);
        }

        if (name == "noSemicolon")
        {
            var noSemi = terms["noSemi"];
            return new ExternalTokenizer((input, stack) =>
            {
                var next = input.Next;
                if (Space.Contains(next)) return;
                if (next == Slash)
                {
                    var after = input.Peek(1);
                    if (after == Slash || after == Star) return;
                }
                if (next != BraceR && next != Semicolon && next != -1 && !((bool?)stack.Context == true))
                    input.AcceptToken(noSemi);
            }, contextual: true);
        }

        if (name == "noSemicolonType")
        {
            var noSemiType = terms["noSemiType"];
            return new ExternalTokenizer((input, stack) =>
            {
                if (input.Next == BracketL && !((bool?)stack.Context == true))
                    input.AcceptToken(noSemiType);
            }, contextual: true);
        }

        if (name == "operatorToken")
        {
            var incdec = terms["incdec"];
            var incdecPrefix = terms["incdecPrefix"];
            var questionDot = terms["questionDot"];
            return new ExternalTokenizer((input, stack) =>
            {
                var next = input.Next;
                if (next == Plus || next == Minus)
                {
                    input.Advance();
                    if (next == input.Next)
                    {
                        input.Advance();
                        var mayPostfix = !((bool?)stack.Context == true) && stack.CanShift(incdec);
                        input.AcceptToken(mayPostfix ? incdec : incdecPrefix);
                    }
                }
                else if (next == Question && input.Peek(1) == Dot)
                {
                    input.Advance();
                    input.Advance();
                    if (input.Next < 48 || input.Next > 57)
                        input.AcceptToken(questionDot);
                }
            }, contextual: true);
        }

        if (name == "jsx")
        {
            var JSXStartTag = terms["JSXStartTag"];
            return new ExternalTokenizer((input, stack) =>
            {
                if (input.Next != Lt) return;
                var parser = stack.Parser;
                var keys = parser.Dialects.Keys.ToArray();
                var jsxIdx = Array.IndexOf(keys, "jsx");
                if (jsxIdx >= 0)
                {
                    if (!stack.DialectEnabled(jsxIdx)) return;
                }
                else return;

                input.Advance();
                if (input.Next == Slash) return;
                var back = 0;
                while (Space.Contains(input.Next))
                {
                    input.Advance(); back++;
                }
                if (JsIdentifierChar(input.Next, true))
                {
                    input.Advance();
                    back++;
                    while (JsIdentifierChar(input.Next, false))
                    {
                        input.Advance(); back++;
                    }
                    while (Space.Contains(input.Next))
                    {
                        input.Advance(); back++;
                    }
                    if (input.Next == Comma) return;
                    var extendsStr = "extends";
                    for (var i = 0; i <= extendsStr.Length; i++)
                    {
                        if (i == extendsStr.Length)
                        {
                            if (!JsIdentifierChar(input.Next, true)) return;
                            break;
                        }
                        if (input.Next != extendsStr[i]) break;
                        input.Advance();
                        back++;
                    }
                }
                input.AcceptToken(JSXStartTag, endOffset: -back);
            }, contextual: true);
        }

        throw new ArgumentException($"Unknown JS external tokenizer: {name}");
    }

    private static readonly NodePropSource JsHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["get set async static"] = Tags.ModifierTag,
        ["for while do if else switch try catch finally return throw break continue default case defer"] = Tags.ControlKeyword,
        ["in of await yield void typeof delete instanceof as satisfies"] = Tags.OperatorKeyword,
        ["let var const using function class extends"] = Tags.DefinitionKeyword,
        ["import export from"] = Tags.ModuleKeyword,
        ["with debugger new"] = Tags.Keyword,
        ["TemplateString"] = Tags.Special(Tags.String),
        ["super"] = Tags.Atom,
        ["BooleanLiteral"] = Tags.Bool,
        ["this"] = Tags.Self,
        ["null"] = Tags.Null,
        ["VariableName"] = Tags.VariableName,
        ["CallExpression/VariableName TaggedTemplateExpression/VariableName"] = Tags.Function(Tags.VariableName),
        ["VariableDefinition"] = Tags.Definition(Tags.VariableName),
        ["Label"] = Tags.LabelName,
        ["PropertyName"] = Tags.PropertyName,
        ["PrivatePropertyName"] = Tags.Special(Tags.PropertyName),
        ["CallExpression/MemberExpression/PropertyName"] = Tags.Function(Tags.PropertyName),
        ["FunctionDeclaration/VariableDefinition"] = Tags.Function(Tags.Definition(Tags.VariableName)),
        ["ClassDeclaration/VariableDefinition"] = Tags.Definition(Tags.ClassName),
        ["NewExpression/VariableName"] = Tags.ClassName,
        ["PropertyDefinition"] = Tags.Definition(Tags.PropertyName),
        ["PrivatePropertyDefinition"] = Tags.Definition(Tags.Special(Tags.PropertyName)),
        ["UpdateOp"] = Tags.UpdateOperator,
        ["LineComment Hashbang"] = Tags.LineComment,
        ["BlockComment"] = Tags.BlockComment,
        ["Number"] = Tags.Number,
        ["String"] = Tags.String,
        ["Escape"] = Tags.Escape,
        ["ArithOp"] = Tags.ArithmeticOperator,
        ["LogicOp"] = Tags.LogicOperator,
        ["BitOp"] = Tags.BitwiseOperator,
        ["CompareOp"] = Tags.CompareOperator,
        ["RegExp"] = Tags.Regexp,
        ["Equals"] = Tags.DefinitionOperator,
        ["Arrow"] = Tags.Function(Tags.Punctuation),
        [": Spread"] = Tags.Punctuation,
        ["( )"] = Tags.Paren,
        ["[ ]"] = Tags.SquareBracket,
        ["{ }"] = Tags.Brace,
        ["InterpolationStart InterpolationEnd"] = Tags.Special(Tags.Brace),
        ["."] = Tags.DerefOperator,
        [", ;"] = Tags.Separator,
        ["@"] = Tags.Meta,
        ["TypeName"] = Tags.TypeName,
        ["TypeDefinition"] = Tags.Definition(Tags.TypeName),
        ["type enum interface implements namespace module declare"] = Tags.DefinitionKeyword,
        ["abstract global Privacy readonly override"] = Tags.ModifierTag,
        ["is keyof unique infer asserts"] = Tags.OperatorKeyword,
        ["JSXAttributeValue"] = Tags.AttributeValue,
        ["JSXText"] = Tags.Content,
        ["JSXStartTag JSXStartCloseTag JSXSelfCloseEndTag JSXEndTag"] = Tags.AngleBracket,
        ["JSXIdentifier JSXNamespacedName"] = Tags.TagName,
        ["JSXAttribute/JSXIdentifier JSXAttribute/JSXNamespacedName"] = Tags.AttributeName,
        ["JSXBuiltin/JSXIdentifier"] = Tags.Standard(Tags.TagName),
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var termTable = JavaScriptParserData.TermTable!;
        var externals = new Dictionary<string, ITokenizer>
        {
            ["insertSemicolon"] = MakeJsExternalTokenizer("insertSemicolon", termTable),
            ["noSemicolon"] = MakeJsExternalTokenizer("noSemicolon", termTable),
            ["noSemicolonType"] = MakeJsExternalTokenizer("noSemicolonType", termTable),
            ["operatorToken"] = MakeJsExternalTokenizer("operatorToken", termTable),
            ["jsx"] = MakeJsExternalTokenizer("jsx", termTable),
        };
        var spec = JavaScriptParserData.MakeSpec(
            externals: externals,
            propSources: [JsHighlighting],
            context: JsTrackNewline
        );
        return new LRParser(spec);
    }
}
